### PR TITLE
[FLINK-31773][runtime] Separate DefaultLeaderElectionService.start(LeaderContender) into two separate methods for starting the driver and registering a contender

### PR DESCRIPTION
**Pull request by XComp**
PR order:
- https://github.com/apache/flink/pull/21742
- https://github.com/apache/flink/pull/22379
- https://github.com/apache/flink/pull/22422
- === this PR === FLINK-31773
- https://github.com/apache/flink/pull/22384
- https://github.com/apache/flink/pull/22390 (FLINK-31785/FLINK-31786)
- https://github.com/apache/flink/pull/22404
- https://github.com/apache/flink/pull/22601

## What is the purpose of the change

This change is about having a clear separation of the `LeaderElectionDriver` lifecycle and the contender registration. This is necessary because we move the driver initialization into `HighAvailabilityServices` where its lifecycle should be maintained.

## Brief change log

* Introduces a clear separation between the `LeaderElectionDriver`'s lifecycle management and the `LeaderContender` registration in `DefaultLeaderElectionService`
* Moves the driver initialization into the `HighAvailabilityServices`:
  * I didn't put that one in a separate PR because it would have required larger temporary code changes due to the fact that the leader election-related methods of HighAvailabilityServices return `LeaderElectionService` instead of `DefaultLeaderElectionService`. The `startLeaderElectionBackend` is not present in the `LeaderElectionService` interface (it's a specific detail of the `DefaultLeaderElectionService`). A few tests (more specifically [ZooKeeperLeaderRetrievalTest:184](https://github.com/apache/flink/blob/381b626a8d3e3b7426de76d0eefba3eba1cba1ba/flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalTest.java#L184)) on the `HighAvailabilityService` to retrieve a `LeaderElectionService`. In these tests, we wouldn't have been able to call the `startLeaderElectionBackend` within the test code but have to rely on it being called within `HighAvailabilityServices`.

## Verifying this change

* Failing tests are updated
* Additional Preconditions are added to ensure consistent behavior

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class LeaderElectionUtils {
+
+    public static String convertToString(
+            Map<String, LeaderInformation> leaderInformationPerContender) {
+        final Map<String, String> convertedMap =
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionUtils.java:33]: Having to construct another map purely for value string conversions seems wrong.
```
map.entrySet().stream()
  .map(entry -> entry.getKey() + "=" + convertToString(entry.getValue())
  .collect(Collectors.joining(", "))
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.guava30.com.google.common.base.Joiner;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class LeaderElectionUtils {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionUtils.java:29]: missing javadoc

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 3]
```diff
@@ -77,22 +97,44 @@ public DefaultLeaderElectionService(LeaderElectionDriverFactory leaderElectionDr
         this.leaderElectionDriver = null;
 
         this.confirmedLeaderInformation = LeaderInformation.empty();
+    }
 
-        this.running = false;
+    public void startLeaderElectionBackend() throws Exception {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:102]: What I missing that makes this usable? This isn't called anywhere in production?

**Review by zentol — COMMENTED** [review]
Did you forget a commit? I don't see a LeaderElection interface.

**Comment by XComp** [comment]
Sorry for not mentioning it before. This PR is still in draft mode because I ran into test failures (I should have updated the PR's description and title). My initial plan with introducing the `LeaderElection` interface first didn't work. It would have required more changes blowing the PR up and making the overall change hard to review. I try a different approach now where I cut the LeaderElectionService methods into pieces first before introducing the `LeaderElection` interface. I'm going to update the PR's description and title accordingly now

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionUtils.java (hunk 1 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionUtils.java:33]: I'm gonna remove these methods for now. They are not used yet but will be used in a future PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java [hunk 4]
```diff
@@ -69,6 +71,118 @@ void testOnGrantAndRevokeLeadership() throws Exception {
...
+
+            contender.waitForLeader();
+
+            testInstance.stop();
+        }
+    }
+
+    /**
+     * This issue can happen when the shutdown of the contender takes too long and the leadership is
+     * re-acquired in the meantime (see FLINK-29234).
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:133]: I dont understand how this comment relates to the test.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java [hunk 5]
```diff
@@ -69,6 +71,118 @@ void testOnGrantAndRevokeLeadership() throws Exception {
...
+            testInstance.stop();
+        }
+    }
+
+    /**
+     * Test to cover the issue described in FLINK-31814. This test could be removed after
+     * FLINK-31814 is resolved.
+     */
+    @Test
+    void testOnRevokeCallWhileClosingService() throws Exception {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:108]: Is the description in FLINK-31814 based on your WIP branch? Asking since there is neither a `DefaultLeaderElection.close()` or `DefaultLeaderElectionService.close` in the current master...

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 6]
```diff
@@ -37,35 +37,58 @@
...
+     * in a supporting IDE.
+     */
     @GuardedBy("lock")
-    // @Nullable is commented-out to avoid having multiple warnings spread over this class
-    // this.running=true ensures that leaderContender != null
-    private volatile LeaderContender leaderContender;
+    private LeaderContender leaderContender;
 
+    /**
+     * Saves the session ID which was issued by the {@link LeaderElectionDriver} iff the leadership
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:64]:
```suggestion
     * Saves the session ID which was issued by the {@link LeaderElectionDriver} if the leadership
```
Not looking forward to more PRs trying to "fix a type".

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 7]
```diff
@@ -37,35 +37,58 @@
...
+     * <p>{@code @Nullable} isn't used here to avoid having multiple warnings spread over this class
+     * in a supporting IDE.
+     */
     @GuardedBy("lock")
-    private volatile boolean running;
-
-    // @Nullable is commented-out to avoid having multiple warnings spread over this class
-    // this.running=true ensures that leaderContender != null
-    private LeaderElectionDriver leaderElectionDriver;
+    private volatile LeaderElectionDriver leaderElectionDriver;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:91]: why is this one now volatile when everything else isn't?

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 8]
```diff
@@ -77,22 +100,48 @@ public DefaultLeaderElectionService(LeaderElectionDriverFactory leaderElectionDr
...
-                            this, new LeaderElectionFatalErrorHandler());
-            LOG.info("Starting DefaultLeaderElectionService with {}.", leaderElectionDriver);
+
+            LOG.info(
+                    "LeaderContender {} has been registered for {}.",
+                    contender.getDescription(),
+                    leaderElectionDriver);
+
+            if (hasLeadership()) {
+                notifyLeaderContenderOfLeadership();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:143]: Wondering if this should really happen in `start()` :thinking: That means it's a different thread than usually, right? Usually it'd be some HA backend thread, but now it's the main thread?

Why isnt this calling `onGrantLeadership`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 9]
```diff
@@ -145,96 +222,122 @@ public void confirmLeadership(UUID leaderSessionID, String leaderAddress) {
...
-            if (running) {
-                handleLeadershipLoss();
+            // TODO: FLINK-31814 covers adding this Precondition
+            // Preconditions.checkState(issuedLeaderSessionID != null,"The leadership should have
+            // been revoked while having the leadership acquired.");
+
+            final UUID previousSessionID = issuedLeaderSessionID;
+            issuedLeaderSessionID = null;
+
+            if (leaderContender != null) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:309]: Is this branching also related to FLINK-31814?

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 10]
```diff
@@ -257,7 +360,7 @@ public void onLeaderInformationChange(LeaderInformation leaderInformation) {
                 }
             } else {
                 LOG.debug(
-                        "Ignoring change notification since the {} has " + "already been closed.",
+                        "Ignoring change notification since the {} has already been closed.",
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:363]:
```suggestion
                        "Ignoring change notification since the {} has already been stopped.",
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 11]
```diff
@@ -145,96 +222,122 @@ public void confirmLeadership(UUID leaderSessionID, String leaderAddress) {
...
         leaderContender.revokeLeadership();
     }
 
     @Override
     public void onLeaderInformationChange(LeaderInformation leaderInformation) {
         synchronized (lock) {
-            if (running) {
+            if (leaderContender != null) {
                 LOG.trace(
                         "Leader node changed while {} is the leader with session ID {}. New leader information {}.",
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:342]: Maybe we can use `LeaderElectionUtils.convertToString(leaderInformation)` to replace `leaderInformation ` as the format param.

**Review by reswqa — COMMENTED** [review]
Thanks @XComp for this work, I have left some comments, PTAL.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java [hunk 12]
```diff
@@ -772,7 +773,11 @@ void testJobMasterServiceLeadershipRunnerCloseWhenElectionServiceGrantLeaderShip
             closeAsyncCalledTrigger.await();
 
             final CheckedThread grantLeadershipThread =
-                    createCheckedThread(currentLeaderDriver::isLeader);
+                    createCheckedThread(
+                            () -> {
+                                currentLeaderDriver.notLeader();
+                                currentLeaderDriver.isLeader();
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java:779]: hmpf, I remember that I put it there because of some Precondition I have added to the `DefaultLeaderElectionService` to ensure that leadership can only be granted when the leadership wasn't acquired before. I'm pretty sure that it was related to the FLINK-31814 issue. :thinking: It's annoying that I cannot reproduce it anymore. Not sure what bug I had still in the `DefaultLeaderElectionService` which I tried to work around with this change. Anyway, you're right: This change is not necessary. I'm gonna revert it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java [hunk 13]
```diff
@@ -70,22 +77,37 @@ public LeaderInformation getLeaderInformation() {
...
     public void isLeader() {
+        isLeader(FutureUtils.completedVoidFuture());
+    }
+
+    public void isLeader(CompletableFuture<Void> grantLeadershipFuture) {
         synchronized (lock) {
             isLeader.set(true);
-            leaderElectionEventHandler.onGrantLeadership(UUID.randomUUID());
+            grantLeadershipFuture.thenRun(
+                    () -> leaderElectionEventHandler.onGrantLeadership(UUID.randomUUID()));
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java:87]: This is actually simulating what we're already using in `DefaultLeaderElectionService` in combination with the `MultipleComponentLeaderElectionDriver` implementations. They call `MultipleComponentLeaderElectionDriver.Listener.isLeader()` which triggers [DefaultMultipleComponentLeaderElectionService.isLeader()](https://github.com/apache/flink/blob/e3cd3b311c1c8a6a0e0cdc849d7c951ef8beea5c/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java#L188). That method calls `onGrantLeadership` for each `LeaderElectionEventHandler` in DefaultMultipleComponentLeaderElectionService:202](https://github.com/apache/flink/blob/e3cd3b311c1c8a6a0e0cdc849d7c951ef8beea5c/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java#L202). These calls will be executed in a separate executor in [DefaultMultipleComponentLeaderElectionService:225](https://github.com/apache/flink/blob/e3cd3b311c1c8a6a0e0cdc849d7c951ef8beea5c/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java#L225).

The order of the grant/revoke calls (which is the important bit) is ensured by the singleThread executor (see [DefaultMultipleComponentLeaderElectionService:98](https://github.com/apache/flink/blob/e3cd3b311c1c8a6a0e0cdc849d7c951ef8beea5c/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java#L98). For test cases that use this feature, the test implementation has to ensure the right order of calls.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java [hunk 14]
```diff
@@ -69,6 +71,118 @@ void testOnGrantAndRevokeLeadership() throws Exception {
...
+     * re-acquired in the meantime (see FLINK-29234).
+     */
+    @Test
+    void testStopWhileHavingLeadership() throws Exception {
+        final TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory driverFactory =
+                new TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory();
+
+        try (final DefaultLeaderElectionService testInstance =
+                new DefaultLeaderElectionService(driverFactory)) {
+            testInstance.startLeaderElectionBackend();
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:142]: Initially, I thought of putting it in public interface but realized that it's specific to the `DefaultLeaderElectionService` implementation and, therefore, removed it from the `LeaderElectionService` interface. It just didn't cross my mind to call the method in the constructor. But it's a valid point now that's it's out of the interface. I will change it. :+1:

[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:142]: Ah, now I remember why I added this method in the first place. With the `MultipleComponentLeaderElection*` classes we added this circular dependency between the `DefaultLeaderElectionService` and the `DefaultMultipleComponentLeaderElectionService` which calls the `DefaultLeaderElectionService.onGrantLeadership` while registering the service in [DefaultMultipleComponentLeaederElectionService:152](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java#L152). This call will result in accessing the `DefaultLeaderElectionService` instance which is still in instantiation phase.

I created FLINK-31837 to cover this issue in a follow-up task.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java (hunk 4 above)
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:133]: You're right: It doesn't really add much value. I added the test when noticing a test failure in [JobMasterServiceLeadershipRunnerTest.testJobMasterServiceLeadershipRunnerCloseWhenElectionServiceGrantLeaderShip](https://github.com/apache/flink/pull/21137/files#diff-c010bcd788571261ea6c43a17ad7aaea719a64a7b8a080d76939e7dbbc5a4b32). But the test scenario of `testStopWhileHavingLeadership` is actually more general and doesn't require additional JavaDoc pointing to FLINK-29234. I'm going to remove the JavaDoc.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java (hunk 5 above)
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java:108]: I guess, the description still came from the WIP. I updated FLINK-31814's description to match the code base of this PR.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 6 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:64]: with "fix a type" you mean fix a typo? :thinking: The `iff` was actually intended (see [Wiktionary](https://en.wikipedia.org/wiki/If_and_only_if#:~:text=In%20logic%20and%20related%20fields,true%20or%20both%20are%20false.)). But it probably adds too much confusion: I'm gonna replace `iff`

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 7 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:91]: Because I missed it when I cleaned up the volatile keywords of all the fields. :+1:

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 8 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:143]: args, that's a good catch - I haven't thought about it in that way. I was planning to include a `singleThreadExecutor` similarly to what we have in `DefaultMultipleComponentLeaderElectionService` but hoped that it could be done in a separate PR. But it looks like I have to integrate that one in here as well. :roll_eyes: I'm going to work on it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 9 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:309]: No, it's possible that `onRevokeLeadership` is called without a contender being registered because the two `on*Leadership` methods are bound to the `leaderElectionDriver`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java:87]: If thread-safety must be handled by the LeaderElectionEventHandler, then shall we document it as such in the javadocs?

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:91]: Yes, my mistake. Thanks for pointing it out.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 6 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:64]: I meant typo.

I know what `iff` means. Many people don't, and open PRs to change it to `if`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 15]
```diff
@@ -145,96 +222,122 @@ public void confirmLeadership(UUID leaderSessionID, String leaderAddress) {
...
         synchronized (lock) {
-            if (running) {
-                return leaderElectionDriver.hasLeadership()
-                        && leaderSessionId.equals(issuedLeaderSessionID);
+            if (leaderElectionDriver != null) {
+                if (leaderContender != null) {
+                    return hasLeadership() && leaderSessionId.equals(issuedLeaderSessionID);
+                } else {
+                    LOG.debug(
+                            "hasLeadership is called after the LeaderContender was removed, returning false.");
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:245]: valid point. the log message survived the migration from the WIP branch to the PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionTest.java [hunk 16]
```diff
@@ -302,6 +303,8 @@ void testZooKeeperReelectionWithReplacement() throws Exception {
 
                     // stop leader election service = revoke leadership
                     leaderElectionService[index].stop();
+                    leaderElectionService[index].close();
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionTest.java:306]: good catch :+1:

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java (hunk 6 above)
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:64]: Now I get you initial statement :bulb:

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java:87]: That's related to your [other comment](https://github.com/apache/flink/pull/22380#discussion_r1168771139). I'm inclined to halt this PR and create a separate ticket that moves the thread handling into the `DefaultLeaderElectionService`

[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionDriver.java:87]: I created FLINK-31878 to cover this issue.

**Comment by XComp** [comment]
Thanks for the comments, @zentol and @reswqa . I addressed them. @zentol raised some valid concerns about the thread-safety which I missed in this PR. I decided to create FLINK-31878 and cover the thread-safety issue separately before going ahead with this (and the subsequent) PR(s).

**Comment by XComp** [comment]
Rebased to most-recent version of FLINK-31838 (PR #22422)

**Review by zentol — COMMENTED** [review]
Are we maybe missing some start call somewhere?

In one of the tests the job submission never went through, and another HA e2e test is also failing.

**Comment by XComp** [comment]
Yes, I'm still investigating the issue. I created a follow-up issue FLINK-32029 to cover an issue where we actually swallowed an Exception in the error handling.

This PR is (again) work in progress until the test failures are resolved. I'm gonna ping you as soon as it's in a reviewable state.

**Comment by XComp** [comment]
> Are we maybe missing some start call somewhere?
> In one of the tests the job submission never went through, and another HA e2e test is also failing.

Just to have this documented: The problem was (again) the fact that there's a grant call being triggered from within the  `MultipleComponentLeaderElectionDriverAdapter` while instantiating the driver. Initially, I only allowed runnables being pushed to the leader event thread if the driver is initialized. This prevented the initial grant call (if leadership was already acquired) to be handled properly.

I'm gonna reorganize/rebase the branch and make it ready to be reviewed again.

**Comment by XComp** [comment]
This PR is ready to get another pass. I tried to keep the commits in reasonable chunks hoping that it helps to review things. The commits will be squashed in the end.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 17]
```diff
@@ -32,49 +31,73 @@
...
     @GuardedBy("lock")
-    private volatile boolean running;
+    private volatile LeaderElectionDriver leaderElectionDriver;
 
-    // @Nullable is commented-out to avoid having multiple warnings spread over this class
-    // this.running=true ensures that leaderContender != null
-    private LeaderElectionDriver leaderElectionDriver;
-
-    private final ExecutorService leadershipOperationExecutor;
+    @Nullable private ExecutorService leadershipOperationExecutor;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java:100]: Instead of making this nullable, how about checking `isShutdown`?

Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java]: args, good idea - I should have gone through the `ExecutorService` interface for such a check.

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java [hunk 18]
```diff
@@ -270,65 +390,73 @@ public void onLeaderInformationChange(LeaderInformation leaderInformation) {
...
+        } else {
+            LOG.debug(
+                    "Ignoring change notification since the {} has already been stopped.",
+                    leaderElectionDriver);
         }
     }
 
     private void runInLeaderEventThread(Runnable callback) {
-        if (running) {
+        if (leadershipOperationExecutor != null) {
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java]: No it's not: I missed that one (even with passing over the PR last week -.-). I added a lock around the execute call and added the @GuardedBy annotation to the `leadershipOperationExecutor`. Additional, I removed the `volatile` from `leaderElectionDriver`. That seem to have survived previous changes and is not necessary anymore.

File: flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderBase.java [hunk 19]
```diff
@@ -80,6 +84,20 @@ public Throwable getError() {
...
+     */
+    public void throwErrorIfPresent() throws Throwable {
+        if (error != null) {
+            throw error;
+        }
+
+        if (!errorQueue.isEmpty()) {
+            throw errorQueue.poll();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (XComp) [reply flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderBase.java]: fair point. That reduces the diff for `DefaultLeaderElectionServiceTest`. :+1: I wrapped it into an `AssertionError` now.

**Comment by XComp** [comment]
@flinkbot run azure

**Comment by XComp** [comment]
The e2e test failed due to some (temporary) network connectivity issues within the Docker image (see [logs](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=49011&view=logs&j=bea52777-eaf8-5663-8482-18fbc3630e81&t=43ba8ce7-ebbf-57cd-9163-444305d74117&l=10378))

**Review by zentol — APPROVED** [review]

**Comment by XComp** [comment]
@reswqa anything to add from your side? I squashed the changes into a single commit now.

**Comment by XComp** [comment]
https://github.com/apache/flink/commit/76bd8bd7d4b7e4a559cbd9fc5f430e9276821e53) is just a rebase to `master`

**Comment by reswqa** [comment]
> @reswqa anything to add from your side?

Thanks @XComp and @zentol, Looks good from my side 👍.

**Review by reswqa — APPROVED** [review]

**Comment by XComp** [comment]
@flinkbot run azure

**Comment by XComp** [comment]
@flinkbot run azure
